src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+-- consistent ('toBipartite' am) == True
+-- @
+consistent :: (Ord a, Ord b) => AdjacencyMap a b -> Bool
+consistent (BAM lr rl) = internalEdgeList lr == sort (map swap $ internalEdgeList rl)
+
+-- The list of edges of a bipartite adjacency map.
+--
+-- /Note: this function is for internal use only./
+internalEdgeList :: Map.Map a (Set.Set b) -> [(a, b)]
+internalEdgeList lr = [ (u, v) | (u, vs) <- Map.toAscList lr, v <- Set.toAscList vs ]
+
+-- | Check if a graph contains a given edge.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasEdge x y 'empty'                                == False
+-- hasEdge x y ('edge' x y)                           == True
+-- hasEdge 2 3 ('edge' 1 2)                           == False
+-- hasEdge x y ('overlay' g ('edge' x y))               == True
+-- hasEdge 1 2 ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Left 2))) == False
+-- @
+hasEdge :: (Ord a, Ord b) => a -> b -> AdjacencyMap a b -> Bool
+hasEdge u v (BAM m _) = ((Set.member v) <$> (u `Map.lookup` m)) == Just True
+
+-- | Check if a graph contains a given vertex in the left part.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasLeftVertex x 'empty'           == False
+-- hasLeftVertex x ('leftVertex' x)  == True
+-- hasLeftVertex x ('rightVertex' x) == False
+-- hasLeftVertex 1 ('leftVertex' 2)  == False
+-- @
+hasLeftVertex :: Ord a => a -> AdjacencyMap a b -> Bool
+hasLeftVertex x (BAM lr _) = x `Map.member` lr
+
+-- | Check if a graph contains a given vertex in the right part.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasRightVertex x 'empty'           == False
+-- hasRightVertex x ('rightVertex' x) == True
+-- hasRightVertex x ('leftVertex' x)  == False
+-- hasRightVertex 1 ('rightVertex' 2) == False
+-- @
+hasRightVertex :: Ord b => b -> AdjacencyMap a b -> Bool
+hasRightVertex y (BAM _ rl) = y `Map.member` rl
+
+-- | Check if a graph contains a given vertex in the given part.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasVertex x 'empty'                   == False
+-- hasVertex (Right x) ('rightVertex' x) == True
+-- hasVertex (Right x) ('leftVertex' x)  == False
+-- hasVertex (Left 1) ('leftVertex' 2)   == False
+-- hasVertex . Left                    == 'hasLeftVertex'
+-- hasVertex . Right                   == 'hasRightVertex'
+-- @
+hasVertex :: (Ord a, Ord b) => Either a b -> AdjacencyMap a b -> Bool
+hasVertex (Left x)  = hasLeftVertex x
+hasVertex (Right y) = hasRightVertex y
+
+-- | Check if a graph is empty.
+-- Complecity: /O(1)/ time.
+--
+-- @
+-- isEmpty 'empty'                 == True
+-- isEmpty ('overlay' 'empty' 'empty') == True
+-- isEmpty ('vertex' x)            == False
+-- isEmpty                       == (==) 'empty'
+-- @
+isEmpty :: AdjacencyMap a b -> Bool
+isEmpty (BAM lr rl) = Map.null lr && Map.null rl
+
+-- | The number of vertices in the left part in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- leftVertexCount 'empty'         == 0
+-- leftVertexCount 'leftVertex' x  == 1
```

Comment:
I don't think this compiles -- there are missing parentheses. Same below and in `rightVertexCount` examples.

Reply:
And in `vertexCount`.